### PR TITLE
HTTP: Do not override status codes that are not 200

### PR DESCRIPTION
**Pull request by Yonas-net**

### CODE REVIEW

**Comment by Yonas-net** [comment]
Testprotocol:
**Before**: See [here](https://github.com/Icinga/icinga2/pull/8104#issuecomment-737844396).

**After**:
```bash
< HTTP/1.1 404 Not Found
< Server: Icinga/v2.12.0-rc1-54-g198177cc5
< Content-Type: application/json
< Content-Length: 57
< 
{
    "error": 404.0,
    "status": "No objects found."
* Connection #0 to host localhost left intact
}*
```

```bash
< HTTP/1.1 400 Bad Request
< Server: Icinga/v2.12.0-rc1-54-g198177cc5
< Content-Type: application/json
< Content-Length: 246
< 
{
    "results": [
        {
            "code": 400.0,
            "status": "Cannot remove downtime 'ha-master!swap!9353fe71-9c49-44a3-863d-a60be7fa66b7'. It is owned by scheduled downtime object 'ha-master!swap!run-downtime'"
        }
    ]
* Connection #0 to host localhost left intact
}*
```

File: lib/remote/actionshandler.cpp [hunk 1]
```diff
@@ -100,11 +100,33 @@ bool ActionsHandler::HandleRequest(
 	}
 
 	int statusCode = 500;
+	Array statusCodes;
```
Comment (Al2Klimov) [inline lib/remote/actionshandler.cpp:103]: Array? Aren't there better data structures for unique values?

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: lib/remote/actionshandler.cpp (hunk 1 above)
Comment (Yonas-net) [reply lib/remote/actionshandler.cpp:103]: And they are? In any case I have not found a better one.

Comment (Al2Klimov) [reply lib/remote/actionshandler.cpp:103]: `std::set<int>`

Comment (Yonas-net) [reply lib/remote/actionshandler.cpp:103]: But I wanted to use our data structure and not from `std`.

Comment (Al2Klimov) [reply lib/remote/actionshandler.cpp:103]: What for?

Comment (Yonas-net) [reply lib/remote/actionshandler.cpp:103]: Because they are there. What were they built for, if they can be used?

Comment (Al2Klimov) [reply lib/remote/actionshandler.cpp:103]: For their use cases. This one isn't a such.

File: lib/remote/actionshandler.cpp [hunk 2]
```diff
@@ -100,11 +100,35 @@ bool ActionsHandler::HandleRequest(
...
 
 	for (const Dictionary::Ptr& res : results) {
-		if (res->Contains("code") && res->Get("code") >= 200 && res->Get("code") <= 299) {
-			statusCode = 200;
-			break;
+		if (!res->Contains("code")) {
+			continue;
+		}
+
+		statusCodes.insert( res->Get("code"));
```
Comment (Al2Klimov) [inline lib/remote/actionshandler.cpp:110]: Please `s/( res/(res/`

File: lib/remote/actionshandler.cpp [hunk 3]
```diff
@@ -100,11 +100,35 @@ bool ActionsHandler::HandleRequest(
...
+	}
+
+	int length = statusCodes.size();
```
Comment (Al2Klimov) [inline lib/remote/actionshandler.cpp:113]: Please `s/int/auto/` ...

File: lib/remote/actionshandler.cpp [hunk 4]
```diff
@@ -100,11 +100,35 @@ bool ActionsHandler::HandleRequest(
...
+
+	if (length == 1) {
+		statusCode = *statusCodes.begin();
+	} else if (length > 1) {
```
Comment (Al2Klimov) [inline lib/remote/actionshandler.cpp:117]: ... and `s/1/1u/`

File: lib/remote/actionshandler.cpp [hunk 5]
```diff
@@ -100,11 +100,35 @@ bool ActionsHandler::HandleRequest(
...
+		auto iterator = statusCodes.begin();
+
+		for (const auto& error : statusCodes) {
```
Comment (Al2Klimov) [inline lib/remote/actionshandler.cpp:118]: I consider the name `error` misleading.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
What happens if there are the codes 200, 404 and 500?

**Comment by Yonas-net** [comment]
> What happens if there are the codes 200, 404 and 500?

I can't test it but if one of the two status codes 404 and 500 appears first, it is returned as status code.

**Comment by Al2Klimov** [comment]
You **can** test. In a separate C++ program w/ just this new algorithm.

**Comment by Yonas-net** [comment]
Which of the status codes does it have to return if it must be correct?

**Comment by Al2Klimov** [comment]
Have you tested the thing I requested?

**Comment by Yonas-net** [comment]
> Have you tested the thing I requested?

200

**Comment by Al2Klimov** [comment]
And now explain why exactly this happens.

**Comment by Yonas-net** [comment]
Well, because 200 is the first value of the set container!

**Comment by Al2Klimov** [comment]
> Which of the status codes does it have to return if it must be correct?

See https://github.com/Icinga/icinga2/pull/8104#issuecomment-737862076

File: lib/remote/actionshandler.cpp [hunk 6]
```diff
@@ -99,16 +99,40 @@ bool ActionsHandler::HandleRequest(
...
+		statusCodes.insert(res->Get("code"));
+	}
+
+	auto length = statusCodes.size();
+	auto noneOkLength = noneOkStatusCodes.size();
+
+	if (noneOkLength == 1u) {
+		resultStatusCode = *noneOkStatusCodes.begin();
+	} else if (noneOkLength >= 2u) {
+		resultStatusCode = 500;
```
Comment (Al2Klimov) [inline lib/remote/actionshandler.cpp:124]: Isn't this line redundant?

File: lib/remote/actionshandler.cpp [hunk 7]
```diff
@@ -99,16 +99,40 @@ bool ActionsHandler::HandleRequest(
...
+	} else if (length >= 1u && noneOkLength < 1u){
+		for (const auto& statusCode : statusCodes) {
+			if (statusCode >= 200 && statusCode <= 299) {
```
Comment (Al2Klimov) [inline lib/remote/actionshandler.cpp:127]: Isn't this check redundant?

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
What happens if there's only one status code – 202?

**Comment by Yonas-net** [comment]
> What happens if there's only one status code – 202?

Then of course 200 is returned!

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
That's not what I wanted.

**Comment by Yonas-net** [comment]
You do not want to return 200 as status code at all?

**Comment by Al2Klimov** [comment]
Not in that particular case.

File: lib/remote/actionshandler.cpp [hunk 8]
```diff
@@ -99,16 +99,29 @@ bool ActionsHandler::HandleRequest(
 		}
 	}
 
-	int statusCode = 500;
+	int resultStatusCode = 500;
```
Comment (Al2Klimov) [inline lib/remote/actionshandler.cpp:102]: Why did you rename this one?

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
What happens if there are only multiple 2xx ones?

**Comment by Yonas-net** [comment]
> What happens if there are only multiple 2xx ones?

500

**Comment by Al2Klimov** [comment]
Do you consider that logical?

File: lib/remote/actionshandler.cpp [hunk 9]
```diff
@@ -100,12 +100,34 @@ bool ActionsHandler::HandleRequest(
...
+
+	auto length = statusCodes.size();
+	auto noneOkSize = noneOkStatusCodes.size();
+
+	if (length == 1u && noneOkSize == 0) {
+		statusCode = *statusCodes.begin();
+	} else if (noneOkSize == 1u) {
+		statusCode = *noneOkStatusCodes.begin();
+	} else if (noneOkSize >= 2u) {
+		statusCode = 500;
```
Comment (Al2Klimov) [inline lib/remote/actionshandler.cpp:128]: Isn't it already initialized as 500?

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: lib/remote/actionshandler.cpp (hunk 9 above)
Comment (Yonas-net) [reply lib/remote/actionshandler.cpp:128]: Yes, but I need this `esle if` condition so that so it won't jump to the next one although the noneOk status codes are greater than 2.

File: lib/remote/actionshandler.cpp [hunk 10]
```diff
@@ -100,12 +100,32 @@ bool ActionsHandler::HandleRequest(
...
+
+	auto length = statusCodes.size();
+	auto noneOkSize = noneOkStatusCodes.size();
+
+	if (length == 1u && noneOkSize == 0) {
+		statusCode = *statusCodes.begin();
+	} else if (noneOkSize == 1u) {
+		statusCode = *noneOkStatusCodes.begin();
+	} else if (length >= 2u && noneOkSize == 0) {
+		statusCode = 200;
```
Comment (Al2Klimov) [inline lib/remote/actionshandler.cpp:128]: Please `s/== 0/== 0u/`

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
Squash everything.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
Write a test protocol for all of the cases we discovered. To make it easier you _may_ test it in a separate C++ program.

**Comment by Yonas-net** [comment]
Testprotocol 2:

**After**:

``vector<int> vec {202};`` => `202`.
``vector<int> vec {404};`` => `404`.
``vector<int> vec {404, 400};`` => `500`.
``vector<int> vec {202, 203};`` => `200`.
``vector<int> vec {202, 404, 202};`` => `404`.
``vector<int> vec {202, 200, 202};``=> `200`.
``vector<int> vec {404, 200, 500};`` => `500`.

**Review by Al2Klimov — APPROVED** [review]

**Review by julianbrost — CHANGES_REQUESTED** [review]
Let me suggest a slightly different algorithm that should do the same with more concise code:

Put all status codes in one set and remove the second one, then do something like this at the end:

```cpp
int statusCode = 500;
if (statusCode.size() == 1u) {
  statusCode = *statusCodes.begin();
} else if (std::all_of(statusCodes.begin(), statusCodes.end(), [](int code) { return 200 <= code && code < 300; })) {
  statusCode = 200;
}
```

File: lib/remote/actionshandler.cpp [hunk 11]
```diff
@@ -100,12 +100,21 @@ bool ActionsHandler::HandleRequest(
 	}
 
 	int statusCode = 500;
+	std::set<int> statusCodes;
 
 	for (const Dictionary::Ptr& res : results) {
-		if (res->Contains("code") && res->Get("code") >= 200 && res->Get("code") <= 299) {
-			statusCode = 200;
-			break;
+		if (!res->Contains("code")) {
```
Comment (julianbrost) [inline lib/remote/actionshandler.cpp:106]: Now that the body of this loop got quite short, this early continue is of little use. I'd just invert the condition and put the insert within the if.

File: lib/remote/actionshandler.cpp [hunk 12]
```diff
@@ -100,12 +100,21 @@ bool ActionsHandler::HandleRequest(
...
+			continue;
 		}
+
+		statusCodes.insert(res->Get("code"));
+	}
+
+	if (statusCodes.size() == 1u) {
+		statusCode = *statusCodes.begin();
+	} else if (std::all_of(statusCodes.begin(), statusCodes.end(),
+	                       [](int code) { return 200 <= code && code < 300; })) {
```
Comment (julianbrost) [inline lib/remote/actionshandler.cpp:116]: Do we have preferred style for line continuation? I think it's fine but maybe someone else can comment on this.

(I think we really move to using clang-format. Manual nitpicking on coding style just feels wrong)

**Review by julianbrost — CHANGES_REQUESTED** [review]

**Comment by julianbrost** [comment]
@Al2Klimov re-requested a review as the code changed. Maybe you can also comment on coding style.

(Also, I'm unable to press the correct buttons...)

**Review by Al2Klimov — COMMENTED** [review]
@Yonas-net Please re-test https://github.com/Icinga/icinga2/pull/8532#pullrequestreview-545177384 (just to be sure).

**Comment by Yonas-net** [comment]
> @Yonas-net Please re-test [#8532 (review)](https://github.com/Icinga/icinga2/pull/8532#pullrequestreview-545177384) (just to be sure).

I have tested it and works just as it did before.

**Comment by Al2Klimov** [comment]
Also the 5th test case of https://github.com/Icinga/icinga2/pull/8532#issuecomment-739762349 ?

**Comment by Yonas-net** [comment]
No, there comes out status code ``500``.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
Fix it and re-test all cases by yourself.

By the way indent changed code w/ tabs.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
You’ve still some spaces (instead of tabs) in the indentation.

File: lib/remote/actionshandler.cpp [hunk 13]
```diff
@@ -100,12 +100,29 @@ bool ActionsHandler::HandleRequest(
...
-			statusCode = 200;
-			break;
+		if (!res->Contains("code")) {
+			continue;
 		}
+
+		if (res->Get("code") >= 200 && res->Get("code") <= 299) {
+			statusCodes.insert(res->Get("code"));
+		} else {
+			nonOkStatusCodes.insert(res->Get("code"));
```
Comment (Al2Klimov) [inline lib/remote/actionshandler.cpp:113]: Call `res->Get("code")` only once.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
Re-test the already written test protocol.

**Comment by Yonas-net** [comment]
I have already tested them all and everything fits.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
Squash your commits.

**Review by Al2Klimov — APPROVED** [review]

File: lib/remote/actionshandler.cpp [hunk 14]
```diff
@@ -100,12 +100,31 @@ bool ActionsHandler::HandleRequest(
...
-			statusCode = 200;
-			break;
+		if (!res->Contains("code")) {
+			continue;
 		}
+
+		auto getCode = res->Get("code");
+
+		if (getCode >= 200 && getCode <= 299) {
+			statusCodes.insert(getCode);
```
Comment (julianbrost) [inline lib/remote/actionshandler.cpp:113]: Since this does not contain all status codes, I'd call the set `okStatusCodes` instead.

File: lib/remote/actionshandler.cpp [hunk 15]
```diff
@@ -100,12 +100,31 @@ bool ActionsHandler::HandleRequest(
...
 
 	for (const Dictionary::Ptr& res : results) {
-		if (res->Contains("code") && res->Get("code") >= 200 && res->Get("code") <= 299) {
-			statusCode = 200;
-			break;
+		if (!res->Contains("code")) {
+			continue;
 		}
+
+		auto getCode = res->Get("code");
```
Comment (julianbrost) [inline lib/remote/actionshandler.cpp:110]: `getCode` makes little sense to me as a variable name, I'd go for just `code`.

File: lib/remote/actionshandler.cpp [hunk 16]
```diff
@@ -100,12 +100,31 @@ bool ActionsHandler::HandleRequest(
...
+
+		if (getCode >= 200 && getCode <= 299) {
+			statusCodes.insert(getCode);
+		} else {
+			nonOkStatusCodes.insert(getCode);
+		}
+	}
+
+	size_t length = statusCodes.size();
+	size_t nonOkSize = nonOkStatusCodes.size();
```
Comment (julianbrost) [inline lib/remote/actionshandler.cpp:120]: These should have consistent names, I'd suggest renaming `length` to `okSize`.

**Review by julianbrost — CHANGES_REQUESTED** [review]

**Review by julianbrost — APPROVED** [review]
